File: examples/python/echo/handler.py
"""
Echo handler -- returns the input text back.

Mirrors the Node echo example handler.
"""

from __future__ import annotations

import json
from typing import Any, Dict, Optional

from blocks_network.types import StartTaskMessage, TaskContext


def handler(task: StartTaskMessage, ctx: Optional[TaskContext] = None) -> Dict[str, Any]:
    """Echo handler - returns the input text back.

    Extracts text from request parts and returns it as plain text.
    """
    # request_parts is the SDK's structured input — each part carries text or file data
    text = _extract_text(task.request_parts or [])

    if ctx:
        # report_status publishes a real-time progress event to task subscribers
        ctx.report_status("Processing...")

    result = f"Processed: {text}"

    # artifacts is the handler's final output, delivered to the consumer as artifact events
    return {
        "artifacts": [{"data": result, "mimeType": "text/plain"}],
    }


def _extract_text(parts: list) -> str:
    """Extract text from request parts, handling RequestPart, dict, and string formats."""
    for part in parts:
        if isinstance(part, str):
            return part
        # RequestPart dataclass with .text attribute
        if hasattr(part, "text") and isinstance(part.text, str):
            return part.text
        if isinstance(part, dict):
            if isinstance(part.get("text"), str):
                return part["text"]
    if parts:
        first = parts[0]
        if hasattr(first, "to_dict"):
            return json.dumps(first.to_dict())
        return json.dumps(first)
    return "Hello from Blocks!"

File: examples/node/echo/handler.ts
import type { StartTaskMessage, TaskContext, HandlerResult } from '@blocks-network/sdk';

/**
 * Echo handler -- returns the input text back.
 *
 * This is the simplest possible Blocks handler: parse the input,
 * optionally report status, and return a single artifact.
 */
const TOTAL_PROCESSING_MS = 15_000;
const STATUS_INTERVAL_MS = 2_000;

export default async function handler(
  task: StartTaskMessage,
  ctx?: TaskContext,
): Promise<HandlerResult> {
  // requestParts is the SDK's structured input — each part carries text or file data
  const input = task.requestParts?.[0];
  const text = extractText(input);

  const start = Date.now();
  let elapsed = 0;
  while (elapsed + STATUS_INTERVAL_MS < TOTAL_PROCESSING_MS) {
    await sleep(STATUS_INTERVAL_MS);
    elapsed = Date.now() - start;
    const remaining = Math.max(0, Math.ceil((TOTAL_PROCESSING_MS - elapsed) / 1000));
    // reportStatus publishes a real-time progress event to task subscribers
    ctx?.reportStatus(`Processing... ${remaining}s remaining`);
  }
  await sleep(Math.max(0, TOTAL_PROCESSING_MS - (Date.now() - start)));

  const result = `Processed: ${text}`;

  // artifacts is the handler's final output, delivered to the consumer as artifact events
  return {
    artifacts: [{ data: result, mimeType: 'text/plain' }],
  };
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function extractText(input: unknown): string {
  if (input !== null && typeof input === 'object') {
    const part = input as Record<string, unknown>;
    if (typeof part.text === 'string') return part.text;
  }
  if (typeof input === 'string') return input;
  return JSON.stringify(input ?? 'Hello from Blocks!');
}
